### PR TITLE
Quantity price attriubute

### LINKED ISSUE #693: Possibility to add and change the price in some services
In presta MP we had possibility to add the price and change the price depending on the number of users. You can see this functionality here: https://marketplace.egi.eu/fitsm/15-fitsm-foundation-level.html

We would like to have something analogous in RoR MP.

### PR DESCRIPTION
**Pull request by mkasztelnik**
Example configuration:

```
{
  "id"=>"id7",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>100, "start_price"=>1024},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
```

Demo:

bedzie

Fixes #693

### CODE REVIEW

**Comment by agpul** [comment]
I have one comment. How we can realize the case when we have one offer with two price option, as I mentioned in the ticket (for eg. the same offer has different price for EGI participant and different price for others).

**Comment by mkasztelnik** [comment]
@agpul two attributes. One for "EGI Participant" second for "outside people"

**Comment by mkasztelnik** [comment]
You are right, but this is general remark (maybe already solved by current implementation) - @bwilk how optional attributes can be modeled in current attribute implementation?

File: app/models/attribute.rb [hunk 1]
```diff
@@ -114,6 +114,8 @@ def self.from_json(json)
       attr = Attribute::Date.new
     when "range-property"
       attr = Attribute::RangeProperty.new
+    when "quantity_price"
```
Comment (goreck888) [inline app/models/attribute.rb]: That's I mean in my review. People who will update offers will get crazy if we won't be consistent.

Comment (bwilk) [reply app/models/attribute.rb]: Agreed

**Comment by bwilk** [comment]
> You are right, but this is general remark (maybe already solved by current implementation) - @bwilk how optional attributes can be modeled in current attribute implementation?

There is no way to model something _optional value_ on a attribute level (as a generic feature). Consider how would optional _select_ look like? Strange istn't it? So - _optional_ value should be modeled as a part of the specific attribute type - e.g. input

**Comment by roksanaer** [comment]
apart from the attribute type name consistency, this PR is OK for this release. I do not perceive this
>>  I have one comment. How can we realize the case when we have one offer with two price option, as I mentioned in the ticket (for eg. the same offer has different price for EGI participant and different price for others).
as a problem. In this case we have 2 separate attributes with 2 different formulas. Both with a default value for number of participants set to zero.
Hovewer we need to better name and define the attributes meant for the specific use case shat stands behind this PR (FitSM training offers).
@agpul can you define in this PR those attributes (config in presta as a source) basing on the expample Marek gave?

**Comment by agpul** [comment]
1. First one is FitSM [https://marketplace.docker-fid.grid.cyf-kr.edu.pl/services/fitsm](url)

* FitSM Foundation Level:
 - EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>80, "start_price"=>1440},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
 - EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>80, "start_price"=>1680},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
* Advanced Level in Service Planning and Delivery:
- EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>160, "start_price"=>2880},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
 - EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>160, "start_price"=>3360},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}  
* Advanced Level in Service Operation and Control:
- EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>160, "start_price"=>2880},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
 - EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>160, "start_price"=>3360},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}  

* Expert Level:
 - EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>240, "start_price"=>3640},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
 - EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>240, "start_price"=>4240},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
} 

* Expert Bridge: (we don't have such option now but I believe we will have it)
 - EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>400, "start_price"=>4250},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}
 - EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>400, "start_price"=>5000},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
}

**Comment by agpul** [comment]
Second offer is (we don't have it now in MP but we need to add ) ISO 27001:

1. ISO 27001 Foundation:

* EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>150, "start_price"=>4400},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
  }
* EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>150, "start_price"=>5150},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
  }

2. ISO 27001 Professional:

* EGI Participant = yes
  {
  "id"=>"id1",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>250, "start_price"=>8750},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
  }
* EGI Participant = No
  {
  "id"=>"id2",
  "type"=>"quantity_price",
  "label"=>"Price per unit",
  "config"=>{"max"=>15, "currency"=>"EUR", "step_price"=>250, "start_price"=>10250},
  "value_type"=>"integer",
  "description"=>"Choose quantity"
  }

**Comment by agpul** [comment]
I put general attributes here. For sure we (or Mal?) need to add better labels and descriptions for each one.

File: app/models/attribute.rb (hunk 1 above)
Comment (mkasztelnik) [reply app/models/attribute.rb]: Renamed, now configuration should looks like:

**Comment by mkasztelnik** [comment]
Updated according to the review, rebased to latest master. The open question is connected with attribute mandatory/optional issue, but as far as I understand slack conversation this can be solved right now by using default `0` default value.

**Comment by roksanaer** [comment]
>> The open question is connected with attribute mandatory/optional issue, but as far as I understand slack conversation this can be solved right now by using default `0` default value.

true

**Pull request merged**